pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go

```diff
@@ -0,0 +1,106 @@
+package delegate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+
+	oauthapi "github.com/openshift/oauth-apiserver/pkg/oauth/apis/oauth"
+)
+
+// OAuthAccessTokenWatcher implements a Watcher interface to watch OAuthAccessTokens
+// and filter them to match only those for the current user
+type OAuthAccessTokenWatcher struct {
+	wrappedWatcher watch.Interface
+	outgoing       chan watch.Event
+	incoming       <-chan watch.Event
+	stopCh         chan struct{}
+
+	mutex   sync.Mutex
+	stopped bool
+}
+
+var _ watch.Interface = &OAuthAccessTokenWatcher{}
+
+// NewOAuthAccessTokenWatcher creates new OAuthAccessTokenWatcher by wrapping a channel
+func newOAuthAccessTokenWatcher(wrappedWatcher watch.Interface) *OAuthAccessTokenWatcher {
+	return &OAuthAccessTokenWatcher{
+		wrappedWatcher: wrappedWatcher,
+		incoming:       wrappedWatcher.ResultChan(),
+		outgoing:       make(chan watch.Event),
+		stopCh:         make(chan struct{}),
+		stopped:        false,
+	}
+}
+
+// Stop implements Interface
+func (w *OAuthAccessTokenWatcher) Stop() {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	if !w.stopped {
+		w.stopped = true
+		// forward the Stop() request
+		w.wrappedWatcher.Stop()
+		close(w.stopCh)
+	}
+}
+
+// Stopping returns true if Stop() has been called
+func (w *OAuthAccessTokenWatcher) Stopping() bool {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	return w.stopped
+}
+
+// ResultChan returns channel for filtered results
+func (w *OAuthAccessTokenWatcher) ResultChan() <-chan watch.Event {
+	return w.outgoing
+}
+
+// StopChan returns stop channel
+func (w *OAuthAccessTokenWatcher) StopChan() <-chan struct{} {
+	return w.stopCh
+}
+
+func (w *OAuthAccessTokenWatcher) Watch(ctx context.Context) {
+	defer close(w.outgoing)
+	defer utilruntime.HandleCrash()
+
+	for {
+		select {
+		case <-w.stopCh:
+			return
+		case event := <-w.incoming:
+			switch event.Type {
+			case watch.Error:
+				w.outgoing <- event
+			default:
+				tokenOrig, ok := event.Object.(*oauthapi.OAuthAccessToken)
+				if !ok {
+					w.outgoing <- createErrorEvent(errors.NewInternalError(fmt.Errorf("failed to convert incoming object to an OAuthAccessToken type")))
+					continue
+				}
+
+				if !isTokenSha256Token(tokenOrig.Name) {
```

Comment:
filter by name too.

Also, I think there's a first class filter concept.

Reply:
Added the filtering.

Not sure whether there's a better, first class filtering. The channel I'm "borrowing" belongs to another watcher that seems to only allow conventional ListOptions